internal/datamodel/confirm_project_commitments.go

```diff
@@ -35,38 +37,71 @@ var (
 	`)
 )
 
-// CanConfirmNewCommitment returns whether the given commitment request can be
-// confirmed immediately upon creation in the given project.
-func CanConfirmNewCommitment(loc core.AZResourceLocation, projectID db.ProjectID, amount uint64, cluster *core.Cluster, dbi db.Interface) (bool, error) {
-	statsByAZ, err := collectAZAllocationStats(loc.ServiceType, loc.ResourceName, &loc.AvailabilityZone, cluster, dbi)
+// CanAcceptCommitmentChangeRequest returns whether the requested moves and creations
+// within the liquid.CommitmentChangeRequest can be done from capacity perspective.
+func CanAcceptCommitmentChangeRequest(req liquid.CommitmentChangeRequest, serviceType db.ServiceType, cluster *core.Cluster, dbi db.Interface) (bool, error) {
+	var distinctResources = make(map[liquid.ResourceName]struct{})
+	for _, projectCommitmentChangeset := range req.ByProject {
+		for resourceName := range projectCommitmentChangeset.ByResource {
+			distinctResources[resourceName] = struct{}{}
+		}
+	}
+	// internally, we only work with projectIDs, so we have to have a conversion ready
+	projectByUUID, err := db.BuildIndexOfDBResult(
+		dbi,
+		func(project db.Project) liquid.ProjectUUID { return project.UUID },
+		`SELECT * FROM projects WHERE uuid = ANY($1)`,
+		pq.Array(slices.Collect(maps.Keys(req.ByProject))))
 	if err != nil {
-		return false, err
+		return false, fmt.Errorf("while building project index: %w", err)
 	}
-	stats := statsByAZ[loc.AvailabilityZone]
 
-	additions := map[db.ProjectID]uint64{projectID: amount}
-	behavior := cluster.CommitmentBehaviorForResource(loc.ServiceType, loc.ResourceName)
-	logg.Debug("checking CanConfirmNewCommitment in %s/%s/%s: projectID = %d, amount = %d",
-		loc.ServiceType, loc.ResourceName, loc.AvailabilityZone, projectID, amount)
-	return stats.CanAcceptCommitmentChanges(additions, nil, behavior), nil
-}
+	for resourceName := range distinctResources {
+		additions := map[db.ProjectID]uint64{}
+		subtractions := map[db.ProjectID]uint64{}
+		additionSum := uint64(0)
+		subtractionSum := uint64(0)
+		for projectUUID, projectCommitmentChangeset := range req.ByProject {
+			project, exists := projectByUUID[projectUUID]
+			// defense in depth: technically, the request has been validated before, so this does not happen.
+			if !exists {
+				return false, fmt.Errorf("project %s not found in database", projectUUID)
+			}
+			for _, commitment := range projectCommitmentChangeset.ByResource[resourceName].Commitments {
+				newStatus, newExists := commitment.NewStatus.Unpack()
+				oldStatus, oldExists := commitment.OldStatus.Unpack()
+				if newExists && newStatus == liquid.CommitmentStatusConfirmed && (!oldExists || oldStatus != liquid.CommitmentStatusConfirmed) {
```

Comment:
The Unpack can be avoided:
```suggestion
				if Commitment.NewStatus == Some(liquid.CommitmentStatusConfirmed) && oldStatus != Some(liquid.CommitmentStatusConfirmed) {
```
Same below.